### PR TITLE
Added feature: Threading in the Controller

### PR DESCRIPTION
**Pull request by mschottdorf**
As  discussed, ControlModuleSimulator inherits from two classes now:
ControlModuleSimulator(ControlModuleBase, threading.Thread)

An instance is started and stopped with Controller.start() and Controller.stop()

A few examples are in the jupyter-notebook

### CODE REVIEW

File: vent/controller/control_module.py [hunk 1]
```diff
@@ -443,30 +441,35 @@ def run(self):
         # start running
```
Comment (sunnyszy) [inline vent/controller/control_module.py:441]: When is `run()` called? I didn't see the class call it.

File: vent/controller/control_module.py [hunk 2]
```diff
@@ -238,14 +232,18 @@ def update(self, pressure):
...
+
+        ControlModuleBase.__init__(self)
+        threading.Thread.__init__(self)
+        self.threadID        = threadID
+        self.name            = name
 
         self.Balloon = Balloon_Simulator(leak=True, delay=False)
         self.Controller = StateController()
-        self.loop_counter = 0
+        self._running = True
```
Comment (sunnyszy) [inline vent/controller/control_module.py:246]: Can you make `_running` default to be False? I feel the use case would click the start button on the UI to start the behavior. And coordinator can also call the `start` function

File: vent/controller/control_module.py [hunk 3]
```diff
@@ -443,30 +441,35 @@ def run(self):
...
+            Qin = self.Controller.get_Qin()
+            self.Balloon.set_flow(Qin, Qout)
 
     def stop(self):
-        # stop running
-        pass
-
-    def recvUIHeartbeat(self):
-        return time.time()
+        self._running = False
```
Comment (sunnyszy) [inline vent/controller/control_module.py:462]: Does this support repetitive stop and start?

File: vent/controller/control_module.py (hunk 2 above)
Comment (mschottdorf) [reply vent/controller/control_module.py:246]: yep, sure

File: vent/controller/control_module.py (hunk 3 above)
Comment (mschottdorf) [reply vent/controller/control_module.py:462]: no, it doesn't. If we want that functionality, we need multiple start and stop functions:
We would need one one start() to start up the thread, and one start() to start the controller loop. stop() would then stop the controller loop, but not the thread.

Is this the goal?

File: vent/controller/control_module.py (hunk 1 above)
Comment (mschottdorf) [reply vent/controller/control_module.py:441]: run()? This is done by python's threading after start()

Here is the ref: "Once a thread object is created, its activity must be started by calling the thread’s start() method. This invokes the run() method in a separate thread of control... Once the thread’s activity is started, the thread is considered ‘alive’. It stops being alive when its run() method terminates" [https://docs.python.org/3/library/threading.html ]

File: vent/controller/control_module.py (hunk 3 above)
Comment (gdoubleyew) [reply vent/controller/control_module.py:462]: I thought we were going to have the UICoordinator or the Main Process start the thread for the controller. And the controller just has the soft start/stop function for the loop. Why is the controller inheriting from thread?

Comment (mschottdorf) [reply vent/controller/control_module.py:462]: That would make sense. In that case yes, no need for threading in the Controller.

Zhenyu requested the change for internal threading. Not sure why. I'm just the code monkey :-)

File: vent/controller/control_module.py (hunk 2 above)
Comment (mschottdorf) [reply vent/controller/control_module.py]: done

File: vent/controller/control_module.py (hunk 3 above)
Comment (mschottdorf) [reply vent/controller/control_module.py]: I've implemented the solution discussed. The class now starts a thread around the main_loop that can be controlled, repetitively, with start() and stop().

File: vent/controller/control_module.py (hunk 1 above)
Comment (sunnyszy) [reply vent/controller/control_module.py:445]: I see. I didn't know this way of inherit a thread. Thanks for letting me knowing this!

Nits: this way seems better in cases where you are creating a thread class (loading threads). But for our current controller class, not sure this is good because later we may have multiple threads inside this class.

Comment (mschottdorf) [reply vent/controller/control_module.py:445]: Yeah, this is not relevant anymore with the revised design

File: vent/controller/control_module.py [hunk 4]
```diff
@@ -240,12 +233,16 @@ def update(self, pressure):
 
 class ControlModuleSimulator(ControlModuleBase):
     # Implement ControlModuleBase functions
-    def __init__(self):
-        super().__init__()  # get all from parent
+    def __init__(self, threadID, name):
+
+        ControlModuleBase.__init__(self)
+        threading.Thread.__init__(self)
```
Comment (sunnyszy) [inline vent/controller/control_module.py:239]: Maybe you also need to clean up these threading base class related code?

File: vent/controller/control_module.py [hunk 5]
```diff
@@ -439,34 +441,50 @@ def get_control(self, control_setting_name: ControlSettingName) -> ControlSettin
...
+        if self._running:
+            print("Controller running...")
+        else:
+            print("Controller not running...")
+        print("Current loop = " + str(self.loop_counter)+'.\n')
 
 def get_control_module(sim_mode=False):
     if sim_mode == True:
-        return ControlModuleSimulator()
+        return ControlModuleSimulator(1, "Controller")
```
Comment (sunnyszy) [inline vent/controller/control_module.py:488]: Also here need clean up

File: vent/controller/control_module.py [hunk 6]
```diff
@@ -347,8 +348,9 @@ def get_sensors(self):
                                           vte=self.vte,
                                           breaths_per_minute=self.bpm,
                                           inspiration_time_sec=self.I_phase,
-                                          timestamp=time.time())
-
+                                          timestamp=time.time(),
+                                          loop_counter = self.loop_counter)
```
Comment (sunnyszy) [inline vent/controller/control_module.py:352]: These variables (updated by main_loop) are not thread safe. So you might need to add a lock when access them. Also rename them to shard_xxx

File: vent/controller/control_module.py (hunk 4 above)
Comment (mschottdorf) [reply vent/controller/control_module.py]: done

File: vent/controller/control_module.py (hunk 5 above)
Comment (mschottdorf) [reply vent/controller/control_module.py]: done

File: vent/controller/control_module.py (hunk 6 above)
Comment (mschottdorf) [reply vent/controller/control_module.py]: added a lock. I'm not entirely sure though, which variables are shared and which are not.

**Comment by mschottdorf** [comment]
This should add a decent fraction of the things discussed today. Importantly, most of the control-code went into the Base-Class. The only two methods which, I believe, cannot be identical between simulation and reality are get_sensors() and start_mainloop().

**Comment by gdoubleyew** [comment]
There's many changes here so it may be hard to follow the thread locking logic. Perhaps we should keep this change as the reorganization and then do a follow on to make everything thread safe. Basically every get_ or set_ interface has to have an internal lock, and each place those variables are accessed other than the get_ or set_ also need to have a lock.

**Comment by sunnyszy** [comment]
I merged the change. And let's deal with the thread locking logic later. Next time I suggest don't make a pull request contains multiple features. For example, this one should only contain internal threading modification.
